File: app/src/main/java/mega/privacy/android/app/fragments/settingsFragments/DownloadSettingsFragment.java
package mega.privacy.android.app.fragments.settingsFragments;

import static mega.privacy.android.app.constants.SettingsConstants.KEY_STORAGE_ASK_ME_ALWAYS;
import static mega.privacy.android.app.constants.SettingsConstants.KEY_STORAGE_DOWNLOAD_LOCATION;
import static mega.privacy.android.app.utils.Constants.REQUEST_DOWNLOAD_FOLDER;
import static mega.privacy.android.app.utils.FileUtil.buildDefaultDownloadDir;
import static mega.privacy.android.app.utils.TextUtil.isTextEmpty;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;

import androidx.preference.Preference;

import java.io.File;

import mega.privacy.android.app.R;
import mega.privacy.android.app.components.TwoLineCheckPreference;
import mega.privacy.android.app.main.FileStorageActivity;
import timber.log.Timber;

public class DownloadSettingsFragment extends SettingsBaseFragment {

    private Preference downloadLocation;
    private TwoLineCheckPreference storageAskMeAlways;

    private boolean askMe;
    private String downloadLocationPath = "";

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        addPreferencesFromResource(R.xml.preferences_download);

        downloadLocation = findPreference(KEY_STORAGE_DOWNLOAD_LOCATION);
        downloadLocation.setOnPreferenceClickListener(this);

        storageAskMeAlways = findPreference(KEY_STORAGE_ASK_ME_ALWAYS);
        storageAskMeAlways.setOnPreferenceClickListener(this);

        if (prefs.getStorageAskAlways() == null) {
            askMe = true;
            dbH.setStorageAskAlways(true);
        } else {
            askMe = Boolean.parseBoolean(prefs.getStorageAskAlways());
        }

        storageAskMeAlways.setChecked(askMe);
        if (askMe) {
            getPreferenceScreen().removePreference(downloadLocation);
        }

        if (isTextEmpty(prefs.getStorageDownloadLocation())) {
            resetDefaultDownloadLocation();
        } else {
            downloadLocationPath = prefs.getStorageDownloadLocation();
        }

        setDownloadLocation();
    }

    private void resetDefaultDownloadLocation() {
        File defaultDownloadLocation = buildDefaultDownloadDir(context);
        downloadLocationPath = defaultDownloadLocation.getAbsolutePath();
    }

    @Override
    public boolean onPreferenceClick(Preference preference) {

        switch (preference.getKey()) {
            case KEY_STORAGE_ASK_ME_ALWAYS:
                dbH.setStorageAskAlways(askMe = storageAskMeAlways.isChecked());
                if (askMe) {
                    getPreferenceScreen().removePreference(downloadLocation);
                } else {
                    resetDefaultDownloadLocation();
                    getPreferenceScreen().addPreference(downloadLocation);
                }
                setDownloadLocation();
                break;

            case KEY_STORAGE_DOWNLOAD_LOCATION:
                toSelectFolder();
                break;
        }

        return true;
    }

    private void setDownloadLocation() {
        dbH.setStorageDownloadLocation(downloadLocationPath);
        prefs.setStorageDownloadLocation(downloadLocationPath);

        if (downloadLocation != null) {
            downloadLocation.setSummary(downloadLocationPath);
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent intent) {
        if (requestCode == REQUEST_DOWNLOAD_FOLDER && intent != null) {
            if (resultCode == Activity.RESULT_OK) {
                downloadLocationPath = intent.getStringExtra(FileStorageActivity.EXTRA_PATH);
                setDownloadLocation();
            } else if (resultCode == Activity.RESULT_CANCELED) {
                Timber.d("REQUEST_DOWNLOAD_FOLDER - canceled");
            }
        }
    }

    private void toSelectFolder() {
        Intent intent = new Intent(context, FileStorageActivity.class);
        intent.setAction(FileStorageActivity.Mode.PICK_FOLDER.getAction());
        intent.putExtra(FileStorageActivity.PICK_FOLDER_TYPE, FileStorageActivity.PickFolderType.DOWNLOAD_FOLDER.getFolderType());
        startActivityForResult(intent, REQUEST_DOWNLOAD_FOLDER);
    }
}

File: app/src/main/java/mega/privacy/android/app/fragments/settingsFragments/DownloadSettingsFragment.kt
package mega.privacy.android.app.fragments.settingsFragments

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import androidx.preference.Preference
import mega.privacy.android.app.R
import mega.privacy.android.app.components.TwoLineCheckPreference
import mega.privacy.android.app.constants.SettingsConstants.KEY_STORAGE_ASK_ME_ALWAYS
import mega.privacy.android.app.constants.SettingsConstants.KEY_STORAGE_DOWNLOAD_LOCATION
import mega.privacy.android.app.main.FileStorageActivity
import mega.privacy.android.app.utils.Constants
import mega.privacy.android.app.utils.FileUtil
import mega.privacy.android.app.utils.TextUtil
import timber.log.Timber

/**
 * The fragment for download settings
 */
class DownloadSettingsFragment : SettingsBaseFragment() {
    private var downloadLocation: Preference? = null
    private var storageAskMeAlways: TwoLineCheckPreference? = null
    private var askMe = false
    private var downloadLocationPath: String? = null

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        addPreferencesFromResource(R.xml.preferences_download)
        downloadLocation = findPreference(KEY_STORAGE_DOWNLOAD_LOCATION)
        storageAskMeAlways = findPreference(KEY_STORAGE_ASK_ME_ALWAYS)

        downloadLocation?.onPreferenceClickListener = this
        storageAskMeAlways?.onPreferenceClickListener = this
        if (prefs.storageAskAlways == null) {
            askMe = true
            dbH.setStorageAskAlways(true)
        } else {
            askMe = java.lang.Boolean.parseBoolean(prefs.storageAskAlways)
        }
        storageAskMeAlways?.isChecked = askMe
        if (askMe) {
            downloadLocation?.let {
                preferenceScreen.removePreference(it)
            }
        }
        if (TextUtil.isTextEmpty(prefs.storageDownloadLocation)) {
            resetDefaultDownloadLocation()
        } else {
            downloadLocationPath = prefs.storageDownloadLocation
        }
        setDownloadLocation()
    }

    private fun resetDefaultDownloadLocation() {
        val defaultDownloadLocation = FileUtil.buildDefaultDownloadDir(context)
        downloadLocationPath = defaultDownloadLocation.absolutePath
    }

    override fun onPreferenceClick(preference: Preference): Boolean {
        when (preference.key) {
            KEY_STORAGE_ASK_ME_ALWAYS -> {
                storageAskMeAlways?.let {
                    dbH.setStorageAskAlways(it.isChecked.also { isChecked -> askMe = isChecked })
                }
                downloadLocation?.let {
                    if (askMe) {
                        preferenceScreen.removePreference(it)
                    } else {
                        resetDefaultDownloadLocation()
                        preferenceScreen.addPreference(it)
                    }
                }
                setDownloadLocation()
            }
            KEY_STORAGE_DOWNLOAD_LOCATION -> toSelectFolder()
        }
        return true
    }

    private fun setDownloadLocation() {
        dbH.setStorageDownloadLocation(downloadLocationPath)
        prefs.storageDownloadLocation = downloadLocationPath
        downloadLocation?.run {
            summary = downloadLocationPath
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, intent: Intent?) {
        if (requestCode == Constants.REQUEST_DOWNLOAD_FOLDER && intent != null) {
            if (resultCode == Activity.RESULT_OK) {
                downloadLocationPath = intent.getStringExtra(FileStorageActivity.EXTRA_PATH)
                setDownloadLocation()
            } else if (resultCode == Activity.RESULT_CANCELED) {
                Timber.d("REQUEST_DOWNLOAD_FOLDER - canceled")
            }
        }
    }

    private fun toSelectFolder() {
        val intent = Intent(context, FileStorageActivity::class.java)
        intent.action = FileStorageActivity.Mode.PICK_FOLDER.action
        intent.putExtra(FileStorageActivity.PICK_FOLDER_TYPE,
            FileStorageActivity.PickFolderType.DOWNLOAD_FOLDER.folderType)
        startActivityForResult(intent, Constants.REQUEST_DOWNLOAD_FOLDER)
    }
}